### PR TITLE
cleanup (collections): Clean up examples, types and file structure

### PR DESCRIPTION
**Pull request by LionC**
This does several tedious clean up tasks in the `collections` module:

- Unify all examples
    - Use `ts` tag for markdown code
    - Use `assertEquals` instead of `console.assert` to have actually working examples
- Move tests into their own subfolder, as the actual folder is being browsed by users and got
quite crowded
- Flatten argument types. Even though `Predicate` and `Selector` read nicely, their meaning
is not immediately clear and LSP hovers do not show their content. The types in the module are
mostly straightforward anyways

### CODE REVIEW

**Comment by kt3k** [comment]
@LionC Thank you for the clean-ups!

I mostly agree with these changes, but I'm not in favor of moving test files into their own folder. Other std modules put the test files next to the implementation, and in my view that flat structure is easier for the collaborators to work with because the corresponding implementations and tests are close to each other.

**Comment by LionC** [comment]
> @LionC Thank you for the clean-ups!
> 
> 
> 
> I mostly agree with these changes, but I'm not in favor of moving test files into their own folder. Other std modules put the test files next to the implementation, and in my view that flat structure is easier for the collaborators to work with because the corresponding implementations and tests are close to each other.

The only problem I see with the current structure is that we explicitly allow `import`ing single functions from this folder, which means users of the module need to navigate it.

I work on the module a lot right now, and even I have a hard time to reorientate myself at a glance when opening the folder.ä, there is just a lot of files with similiar names directly next to each other, some being tests and some being implementations without a consistent ordering pattern (because of the default alphabetical sorting everywhere).

My opinion on this is not super steong, but I can say that working on the module and scanning the folders at a glance is already a lot nicer after moving the test files. I will leave final judgement to you and the others though :-)

**Comment by caspervonb** [comment]
>  I will leave final judgement to you and the others though :-)

Test module next to the module being has been the way since the start and is the convention everywhere else so let's not randomly break from that for `collections`.

**Comment by LionC** [comment]
Done, please re-review @kt3k

File: collections/group_by.ts [hunk 1]
```diff
@@ -22,16 +21,16 @@ import { Grouping, Selector } from "./types.ts";
...
  *     'K': [ { name: 'Kim' } ],
  * })
  * ```
  */
 export function groupBy<T>(
   array: Array<T>,
-  selector: Selector<T, string>,
-): Grouping<T> {
+  selector: (el: T) => string,
+): { [key: string]: Array<T> } {
```
Comment (zhangyuannie) [inline collections/group_by.ts:33]: Minor nitpick, we might want to use `Record` here since that is what's used throughout `collections`.

Comment (LionC) [reply collections/group_by.ts:33]: You are right

Comment (LionC) [reply collections/group_by.ts]: Done.

**Review by kt3k — APPROVED** [review]
@LionC Thanks! LGTM

**Pull request merged**